File: Ruby/Collatz.rb
#!/usr/bin/env ruby
#Author: Jarrett Melnick
#Program to print numbers with the smallest Collatz values in a given range.
class CollatzNumber
	def initialize(identity, sequence)
		@identity, @sequence = identity, sequence
	end
	def GetIdentity
		@identity
	end
	def GetSequence
		@sequence
	end
end
def Collatz(num = 1)
	#Function to perform the collatz sequence on a given number
	seq = 0
	until num < 2 		#Essentially while(!true). Will run until the condition is met
		num = (num.to_s(2)[-1] == '0') ? (num/2) : (3*num+1)
		seq += 1
	end
	return seq
end
def GetLargest(numArray, seq)
	#Returns the index of the number with the largest collatz sequence
	if seq == 1
		max = 0
		numArray.each_index{|i|
			if numArray[i] != nil
				if numArray[max].GetSequence < numArray[i].GetSequence
					max = i
				end
			end
		}
	else
		max = 0
		numArray.each_index{|i|
			if numArray[i] != nil
				if numArray[max].GetIdentity < numArray[i].GetIdentity
					max = i
				end
			end
		}
	end
	return max
end
def GetSmallest(numArray)
	#Returns the index of the number wiht the smallest collatz sequence
	min = 0;
	numArray.each_index{|i|
		if numArray[i] != nil
			if numArray[min].GetSequence > numArray[i].GetSequence
				min = i
			end
		end
	}
	return min
end
def HasCollatzNumber(numArray, val)
	#Checks if the array contains val. Returns true if val is found.
	numArray.each_index{|i|
		if numArray[i] != nil
			if numArray[i].GetSequence == val
				return true
			end
		end
	}
	return false
end

print "Please provide an endpoint: "
iterate = gets.to_i			#The largest number to be considered. Changes range of numbers iterated.

nums = Array.new(10, CollatzNumber.new(0,0))
nums2 = Array.new(10, CollatzNumber.new(0,0))

puts "Calculating Collatz Values..."
for i in 2..iterate
	smallest = GetSmallest(nums)
	curCollatz = Collatz(i)
	if((curCollatz > nums[smallest].GetSequence) and !(HasCollatzNumber(nums, curCollatz)))
		num = CollatzNumber.new(i, curCollatz)
		nums[smallest] = num
		nums2[smallest] = num
	end
end
puts "SORTED BY SEQUENCE LENGTH: "
#Prints top 10 numbers with the largest collatz sequence.
for i in 1..10
	largest = GetLargest(nums,1)
	puts "#{nums[largest].GetIdentity} has a collaz sequence of #{nums[largest].GetSequence}"
	nums[largest] = CollatzNumber.new(0,0)
end
puts "\nSORTED BY INTEGER LENGTH:"
for i in 1..10
	largest = GetLargest(nums2, 0)
	puts "#{nums2[largest].GetIdentity} has a collatz sequence of #{nums2[largest].GetSequence}"
	nums2[largest] = CollatzNumber.new(0,0)
end

File: Python/Collatz.py
#!/usr/bin/python
#Author: Jarrett Melnick
#Program to calculate the top 10 largest Collatz values for a given range.
class CollatzNumber:
	def __init__(self, identity,sequence):
		self.identity = long(identity)
		self.sequence = long(sequence)

def Collatz(num):
	#Function to calculate the length the Collatz sequence for the given number.
	seq = 0
	while(num > 1):
		n = bin(num)
		if(n[-1] == '1'):
			num = 3*num+1
			num = num/2
			seq = seq + 2
		else:
			num = num/2
			seq = seq + 1

	return seq

def GetLargest(N,seq):
	#Gets the largest number in the dictionary table 
	maxi = 1
	if(seq == 1):
		for i in range(0,10):
			if(N[i].sequence >= N[maxi].sequence):
				maxi = i
		return maxi
	else:
		for i in range(0,10):
			if(N[i].identity >= N[maxi].identity):
				maxi = i
		return maxi

def GetSmallest(N):
	#num = CollatzNumber(0,-1)
	#N.append(num)
	mini = 0
	for i in range(1,10):
		if(N[i].sequence < N[mini].sequence):
			mini = i
	#N.pop(10)
	#print "%d has the smallest collatz sequence of length %d" % (Nums[mini].identity, Nums[mini].sequence)
	return mini

def hasCollatzNumber(N, val):
	for i in range(0,10):
		if(N[i].sequence == val):
			return True
	return False

Nums = []		#List to store Collatz Values of the index.
for i in range(0, 10):
	num = CollatzNumber(0, 0)
	Nums.insert(0,num)

print "Please provide an endpoint"
iterate = long(input())
print "CALCULATING VALUES..."
for i in xrange(2, iterate+1):
	curCollatz = Collatz(i)
	smallest = GetSmallest(Nums)
	if((curCollatz > Nums[smallest].sequence) and not(hasCollatzNumber(Nums,curCollatz))):
		num = CollatzNumber(i, curCollatz)
		Nums[smallest] = num
		#print len(Nums)
	
print "SORTED BY SEQUENCE LENGTH:"
Nums2 = list(Nums)
for i in range(0,10):
#Prints the numbers with the top 10 largest Collatz values
	largest = GetLargest(Nums,1)
  	print "%d has a collatz sequence of length %d" % (Nums[largest].identity, Nums[largest].sequence)
  	num = CollatzNumber(0,0)
  	Nums[largest] = num

print "\nSORTED BY INTEGER LENGTH:"
for i in range(0,10):
	largest = GetLargest(Nums2,0)
	print "%d has a collatz sequene of length %d" % (Nums2[largest].identity, Nums2[largest].sequence)
	num = CollatzNumber(0,0)
	Nums2[largest] = num
